cpp/src/groupby/hash/m2_var_functor.cuh

```diff
@@ -17,161 +17,114 @@
-      target{target},
-      source{source},
-      sum{sum},
-      count{count},
-      ddof{ddof}
-  {
-  }
+}  // namespace
 
-  template <typename Source>
-  __device__ void operator()(column_device_view const&, size_type, size_type) noexcept
-    requires(!is_m2_var_supported<Source>())
-  {
-    CUDF_UNREACHABLE("Invalid source type for std, var aggregation combination.");
-  }
-
-  template <typename Source>
-  __device__ void operator()(column_device_view const& source,
-                             size_type source_index,
-                             size_type target_index) noexcept
-    requires(is_m2_var_supported<Source>())
-  {
-    using Target    = cudf::detail::target_type_t<Source, aggregation::VARIANCE>;
-    using SumType   = cudf::detail::target_type_t<Source, aggregation::SUM>;
-    using CountType = cudf::detail::target_type_t<Source, aggregation::COUNT_VALID>;
-
-    if (source.is_null(source_index)) return;
-    CountType group_size = count.element<CountType>(target_index);
-    if (group_size == 0 or group_size - ddof <= 0) return;
-
-    auto x        = static_cast<Target>(source.element<Source>(source_index));
-    auto mean     = static_cast<Target>(sum.element<SumType>(target_index)) / group_size;
-    Target result = (x - mean) * (x - mean) / (group_size - ddof);
-    cuda::atomic_ref<Target, cuda::thread_scope_device> ref{target.element<Target>(target_index)};
-    ref.fetch_add(result, cuda::std::memory_order_relaxed);
-    // STD sqrt is applied in finalize()
-
-    if (target.is_null(target_index)) { target.set_valid(target_index); }
-  }
+inline void compute_m2(data_type source_type,
+                       mutable_column_view const& target,
+                       column_view const& sum_sqr,
+                       column_view const& sum,
+                       column_view const& count,
+                       rmm::cuda_stream_view stream)
+{
+  type_dispatcher(source_type, m2_functor{}, target, sum_sqr, sum, count, stream);
+}
 
-  __device__ inline void operator()(size_type source_index)
-  {
-    if (row_bitmask == nullptr or cudf::bit_is_set(row_bitmask, source_index)) {
-      auto const target_index = *set.find(source_index);
-
-      auto col         = source;
-      auto source_type = source.type();
-      if (source_type.id() == type_id::DICTIONARY32) {
-        col          = source.child(cudf::dictionary_column_view::keys_column_index);
-        source_type  = col.type();
-        source_index = static_cast<size_type>(source.element<dictionary32>(source_index));
-      }
-
-      type_dispatcher(source_type, *this, col, source_index, target_index);
-    }
-  }
-};
+inline void compute_variance(mutable_column_view const& target,
+                             column_view const& m2,
+                             column_view const& count,
+                             bool* validity,
+                             size_type ddof,
+                             rmm::cuda_stream_view stream)
```

Comment:
Could we move these 2 inline functions to a .cu file instead?

Reply:
Done. We need to keep such functions in a separate place (I've put them in the `aggregation/common` folder) so they can be shared across hash- and sort- groupby, which is the follow up work after this one.